File: 2023/12.py
from utils import get_input_data

data = get_input_data(12).splitlines()
data = [line.split() for line in data]
memo = {}


def count_arrangements(config, damaged_map):
    if config == '':
        return 1 if damaged_map == () else 0
    if damaged_map == ():
        return 0 if '#' in config else 1

    key = (config, damaged_map)
    if key in memo:
        return memo[key]

    result = 0
    if config[0] in '.?':
        result += count_arrangements(config[1:], damaged_map)
    if config[0] in '#?':
        if enough_left(config, damaged_map[0]) and all_broken(config, damaged_map[0]) and next_works(config,
                                                                                                     damaged_map[0]):
            result += count_arrangements(config[damaged_map[0] + 1:], damaged_map[1:])

    memo[key] = result
    return result


def enough_left(config, damaged_index):
    return damaged_index <= len(config)


def all_broken(config, damaged_index):
    return '.' not in config[:damaged_index]


def next_works(config, damaged_index):
    return damaged_index == len(config) or config[damaged_index] != '#'


def unfold(config, damaged_map):
    return '?'.join([config] * 5), damaged_map * 5


def solve_part_one():
    total = 0
    for d in data:
        total += count_arrangements(d[0], tuple(map(int, d[1].split(','))))
    return total


def solve_part_two():
    total = 0
    for d in data:
        config, damaged_map = unfold(d[0], tuple(map(int, d[1].split(','))))
        total += count_arrangements(config, damaged_map)
    return total


print('Part One: %d' % solve_part_one())
print('Part Two: %d' % solve_part_two())

File: 2018/12.js
import {getInputData} from './utils.js';

const data = (await getInputData(12)).trim().split('\n');
const initialState = data[0].replace('initial state: ', '');

function generateRecipe() {
	const recipe = new Set();
	for (const line of data.slice(1)) {
		const [pattern, result] = line.split(' => ');
		if (result === '#') recipe.add(pattern);
	}
	return recipe;
}

function nextGeneration(currentSet, recipe) {
	const start = Math.min(...currentSet) - 3;
	const end = Math.max(...currentSet) + 3;
	const nextSet = new Set();
	for (let i = start; i <= end; i++) {
		const pattern = [-2, -1, 0, 1, 2].map(offset => currentSet.has(i + offset) ? '#' : '.').join('');
		if (recipe.has(pattern)) nextSet.add(i);
	}
	return nextSet;
}

function solvePartOne() {
	const recipe = generateRecipe();
	let currentSet = new Set();
	for (let i = 0; i < initialState.length; i++) if (initialState[i] === '#') currentSet.add(i);
	for (let generation = 0; generation < 20; generation++) currentSet = nextGeneration(currentSet, recipe);
	return Array.from(currentSet).reduce((sum, index) => sum + index, 0);
}

function solvePartTwo() {
	const recipe = generateRecipe();
	let currentSet = new Set();
	for (let i = 0; i < initialState.length; i++) if (initialState[i] === '#') currentSet.add(i);
	let lastSum = 0;
	let growthRate = 0;
	let stableGenerationsCount = 0;
	const requiredStability = 10;
	const maxGenerationsToCheck = 2000; // Arbitrary safety limit
	const targetGeneration = 50000000000;
	for (let generation = 1; generation <= maxGenerationsToCheck; generation++) {
		currentSet = nextGeneration(currentSet, recipe);
		const currentSum = Array.from(currentSet).reduce((sum, index) => sum + index, 0);
		const currentGrowthRate = currentSum - lastSum;
		if (generation > 1) {
			if (currentGrowthRate === growthRate) stableGenerationsCount++;
			else {
				stableGenerationsCount = 0;
				growthRate = currentGrowthRate;
			}
			if (stableGenerationsCount >= requiredStability) {
				const generationsRemaining = targetGeneration - generation;
				return currentSum + generationsRemaining * growthRate;
			}
		}
		lastSum = currentSum;
	}

	return Array.from(currentSet).reduce((sum, index) => sum + index, 0);
}

console.log(`Part One: ${solvePartOne()}`);
console.log(`Part Two: ${solvePartTwo()}`);